### PR TITLE
Update entrypoint to add structural folders and indexes, closes MISC-114

### PR DESCRIPTION
**Pull request by tpetillot**
**Motivation:**
* Versioning references need a current folder of the latest version
* Indexes cascading repository and version parameters

**Modification:**
* Update script to create current folder copy of the versioned folder specified as latest
* Create index files cascading parameters such as repository URL and version

**Result:**
* Versioning works properly when aggregated (e.g: current redirection)
* Repository edition links works as expected

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -38,6 +41,58 @@ if [ -z "$DEBUG" ]; then
...
+  build_unversionned_indexes() {
+    envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR}' < "scripts/unversionned-section-index.md" > "content/$local_dir/_index.md"
+  }
+
+  if [[ -n $version ]]; then
+    build_version_indexes
+  else
+    build_unversionned_indexes
+  fi
+}
```
Comment (Isammoc) [inline entrypoint.sh:94]: As you already defined 2 functions and these functions depends on the presence of one argument, you should directly call the corresponding function.

File: entrypoint.sh [hunk 2]
```diff
@@ -51,15 +106,39 @@ fetch_doc () {
...
+  local branch="$2"
+  local remote_dir="$3"
+  local local_dir="$4"
+  local version="$5"
+  local latest="$6"
+
+  if [[ -n $latest ]]; then
+    fetch_doc "$repo" "$branch" "$remote_dir" "$local_dir"
+    mkdir -p "$local_dir/reference/$version"
+    cp -r "$local_dir/reference/current"/* "$local_dir/reference/$version"
```
Comment (Isammoc) [inline entrypoint.sh:120]: Should not be in fetch_doc...

The structure (and the copy) should not be there.

Comment (tpetillot) [reply entrypoint.sh:120]: I don't agree, fetch documentation is responsible to fetch content in the right folder.
When version isn't last, it's fetch inside `content/<local_dir>/<version>`, when `latest`, it fetch all content and still respect the  `content/<local_dir>/<version>`.

File: entrypoint.sh [hunk 3]
```diff
@@ -38,6 +41,40 @@ if [ -z "$DEBUG" ]; then
...
+  local local_dir="$4"
+  local version="$5"
+  local latest="$6"
+
+  export REPOSITORY="$repo" BRANCH="$branch" REMOTE_DIR="$remote_dir" LOCAL_DIR="$local_dir"
+
+  if [[ -n $version ]]; then
+    cp "scripts/versioned-reference-index.md" "content/$local_dir/reference/_index.md"
+    export VERSION="$version" LATEST="${latest:-false}"
+    envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR} ${VERSION} ${LATEST}' < "scripts/versioned-reference-section-index.md" > "content/$local_dir/reference/$version/_index.md"
```
Comment (Isammoc) [inline entrypoint.sh:68]: Shouldn't we use something like:

```suggestion
    cat > "content/$local_dir/reference/$version/_index.md" << EOF
---
cascade:
  version: "${VERSION}"
  latest: ${LATEST}
  docsRepo:
    url: "https://github.com/gatling/${REPOSITORY}/blob/${BRANCH}/${REMOTE_DIR}"
    rel: "${LOCAL_DIR}"
---
EOF
```

Note: no need to export variables with this.

File: entrypoint.sh [hunk 4]
```diff
@@ -38,6 +41,40 @@ if [ -z "$DEBUG" ]; then
...
+  local repo="$1"
+  local branch="$2"
+  local remote_dir="$3"
+  local local_dir="$4"
+  local version="$5"
+  local latest="$6"
+
+  export REPOSITORY="$repo" BRANCH="$branch" REMOTE_DIR="$remote_dir" LOCAL_DIR="$local_dir"
+
+  if [[ -n $version ]]; then
```
Comment (Isammoc) [inline entrypoint.sh:65]: You misunstood my last comment.

Instead having a function with an optional parameter, we should have 2 distinct functions that we directly call.

File: entrypoint.sh (hunk 3 above)
Comment (Isammoc) [reply entrypoint.sh:68]: Ok, I understand your point.
I'm not a big fan of the `envsubst` command either :stuck_out_tongue: 

So we should find a better name for the "`scripts`" directory, then.

What will be the solution for the left menu?
Perhaps having something that manage all this yaml stuff in the same place (and not in the bash script) can be another solution.

**Comment by tpetillot** [comment]
> fetch_doc was responsible to fetch other git repositories.
> In a mono repository, the content will be already there and we still have to copy the current version.

Current is always fetched. If we were going with your point, I would have to drop `$local_dir/reference/current` after the `cp`, and then later copy it back...

File: entrypoint.sh (hunk 2 above)
Comment (tpetillot) [reply entrypoint.sh:109]: Current is always fetched. If we were going with your point, I would have to drop $local_dir/reference/current after the cp, and then later copy it back...

Comment (Isammoc) [reply entrypoint.sh:109]: Huh?

`frontline-doc#main:/content/` contains the current folder with actual current version
then, we will have to copy `frontline-doc#main:/content/reference/current`to `/content/self-hosted/reference/1.14`
No matter if we got it from monorepo or multi repo.

If it is still not clear, let's speak together.

File: entrypoint.sh (hunk 3 above)
Comment (tpetillot) [reply entrypoint.sh]: Menu yaml is now inferred from the structure of the content

Comment (tpetillot) [reply entrypoint.sh]: `template/indexes`

File: bin/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+PORT=1313
+ENVIRONMENT=development
+COMMAND="server"
+WITHDRAFTS=2
+BASEURL=""
+PREPARE=true
+
+FRONTLINE_VERSIONS=("1.13" "1.14")
+FRONTLINE_VERSION_LATEST="1.14"
```
Comment (Isammoc) [inline bin/entrypoint.sh:32]: Not used?

File: bin/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+build_indexes() {
+  local repo="$1"
+  local branch="$2"
+  local remote_dir="$3"
+  local local_dir="$4"
+
+  export REPOSITORY="$repo" BRANCH="$branch" REMOTE_DIR="$remote_dir" LOCAL_DIR="$local_dir"
+  envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR}' < "template/indexes/unversionned-section-index.md" > "content/$local_dir/_index.md"
```
Comment (Isammoc) [inline bin/entrypoint.sh:62]:
```suggestion
 REPOSITORY="$repo" \
   BRANCH="$branch" \
   REMOTE_DIR="$remote_dir" \
   LOCAL_DIR="$local_dir" \
   envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR}' < "template/indexes/unversionned-section-index.md" > "content/$local_dir/_index.md"
```

Shouldn't need to export

File: bin/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  local branch="$2"
+  local remote_dir="$3"
+  local local_dir="$4"
+  local version="$5"
+  local latest="$6"
+
+  export REPOSITORY="$repo" BRANCH="$branch" REMOTE_DIR="$remote_dir" LOCAL_DIR="$local_dir" VERSION="$version" LATEST="${latest:-false}"
+
+  cp "template/indexes/versioned-reference-index.md" "content/$local_dir/reference/_index.md"
+  export VERSION="$version" LATEST="${latest:-false}"
```
Comment (Isammoc) [inline bin/entrypoint.sh:76]: `VERSION` and `LATEST` are exported twice.

File: bin/entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+  envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR} ${VERSION} ${LATEST}' < "template/indexes/versioned-reference-section-index.md" > "content/$local_dir/reference/$version/_index.md"
+
+  if [[ -n $latest ]]; then
+    envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR}' < "template/indexes/unversionned-section-index.md" > "content/$local_dir/_index.md"
+    cp "content/$local_dir/reference/$version/_index.md" "content/$local_dir/reference/current/_index.md"
+  fi
```
Comment (Isammoc) [inline bin/entrypoint.sh:82]:
```suggestion
  cp "template/indexes/versioned-reference-index.md" "content/$local_dir/reference/_index.md"
  REPOSITORY="$repo" \
    BRANCH="$branch" \
    REMOTE_DIR="$remote_dir" \
    LOCAL_DIR="$local_dir" \
    VERSION="$version" \
    LATEST="${latest:-false}" \
    envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR} ${VERSION} ${LATEST}' < "template/indexes/versioned-reference-section-index.md" > "content/$local_dir/reference/$version/_index.md"

  if [[ -n $latest ]]; then
    REPOSITORY="$repo" \
      BRANCH="$branch" \
      REMOTE_DIR="$remote_dir" \
      LOCAL_DIR="$local_dir" \
      envsubst '${REPOSITORY} ${BRANCH} ${REMOTE_DIR} ${LOCAL_DIR}' < "template/indexes/unversionned-section-index.md" > "content/$local_dir/_index.md"
    cp "content/$local_dir/reference/$version/_index.md" "content/$local_dir/reference/current/_index.md"
  fi
```

**Review by Isammoc — APPROVED** [review]

**Pull request merged**